File: Deadlock-detection/src/main/java/CycleDetectionDFS.java
import java.util.*;

public class CycleDetectionDFS {
    int nodes;
    List<List<Integer>> adjList;
    boolean[] recTrack;
    boolean[] visited;
    Map<Integer,Integer> parent;
    List<Integer> cyclePath;
    Scanner sc;
    public CycleDetectionDFS(){
        sc=new Scanner(System.in);
        this.getNumberNodes();
        this.getAdjList();
        this.runAlgo();
    }
    private void getAdjList(){
        this.adjList=new ArrayList<>();
        for(int i=0;i<nodes;i++){
            System.out.printf("Enter the nodes on which process P%d waits for%n",i);
            List<Integer> temp=new ArrayList<>();
            for(int j=0;j<nodes;j++){
                if(i!=j){
                    System.out.printf("If P%d waits on P%d enter 1 else 0 : ",i,j);
                    int val=sc.nextInt();
                    if(val==1){
                        temp.add(j);
                    }
                }
            }
            this.adjList.add(temp);
        }
    }
    private void getNumberNodes(){
        System.out.print("Enter the number of process in the system : ");
        this.nodes=sc.nextInt();
        System.out.println();
    }
    private void initialize(){
        this.recTrack=new boolean[nodes];
        this.visited=new boolean[nodes];
        this.parent=new HashMap<>();
        this.cyclePath=new ArrayList<>();
    }
    public void runAlgo(){
        this.cyclePath=findCycle();
        if(cyclePath==null || cyclePath.isEmpty()){
            return;
        }
        else{
            this.displayCycle(this.cyclePath);
        }
    }
    void displayCycle(List<Integer> path){
        int i=0;
        for(int a:path){
            if(i==0){
                System.out.printf("%d",a);
            }
            else {
                System.out.printf("->%d", a);
            }
            i++;
        }
    }
    public List<Integer> findCycle(){
        this.initialize();
        for(int i=0;i<this.nodes;i++){
            if(!this.visited[i]){
                if(dfs(i)){
                    System.out.println("Cycle has been found and hence deadlock detected!!");
                    return this.cyclePath.reversed();
                }
            }
        }
        System.out.println("No cycle and hence no deadlock has been detected!!");
        return null;
    }

    private boolean dfs(int node){
        this.visited[node]=true;
        this.recTrack[node]=true;
        for(int a:this.adjList.get(node)){
            if(!visited[a]){
                this.parent.put(a,node);
                if(dfs(a)){
                    return true;
                }
            }
            else if(recTrack[a]){
                int curr=node;
                this.cyclePath.add(curr);
                while(curr!=a){
                    curr=this.parent.get(curr);
                    this.cyclePath.add(curr);
                }
                this.cyclePath.add(node);
                return true;
            }
        }
        recTrack[node]=false;
        return false;
    }
}

File: Deadlock-detection-py/CycleDetectionDFS.py
import networkx as nx
import matplotlib.pyplot as plt
class CycleDetectionDFS:
    def __init__(self):
        self.nodes=self.getNumberNodes()
        self.adjList=self.getAdjList()
        self.visited=[False]*self.nodes
        self.recTrack=[False]*self.nodes
        self.cyclePath=[]
        self.parent={}
        
    def getNumberNodes(self):
        self.nodes=int(input("Enter the number of nodes:"))
        return self.nodes
    def getAdjList(self):
        self.adjList=[]
        for i in range(self.nodes):
            print(f"Enter the nodes on which process P{i} waits for:",end="\n")
            temp=[]
            while True:
                node=int(input())
                if node==-1:
                    break
                temp.append(node)
            
            self.adjList.append(temp)
        return self.adjList
    
        
    def findCycle(self):
        for i in range(self.nodes):
            if not self.visited[i]:
                if self.dfs(i):
                    print("Cycle has been found and deadlock is detected!!")
                    
                    self.cyclePath.reverse()
                    return self.cyclePath
        print("No cycle has been found and deadlock is not detected!!")
        return None
    
    def dfs(self,node):
        self.visited[node]=True
        self.recTrack[node]=True
        
        for i in self.adjList[node]:
            if not self.visited[i]:
                self.parent[i]=node
                if self.dfs(i):
                    return True
            elif self.recTrack[i]:
                curr=node
                self.cyclePath.append(i)
                self.cyclePath.append(curr)
                while(curr!=i):
                    curr=self.parent[curr]
                    self.cyclePath.append(curr)
                
                
                return True
        self.recTrack[node]=False
        return False
    def runAlgo(self):
        self.cyclePath=self.findCycle()
        print(self.cyclePath)
        if self.cyclePath:
            self.displayCyclePath()

    def displayCyclePath(self):
        print("Cycle Path is:")
        for i in range(len(self.cyclePath)):
            if i==len(self.cyclePath)-1:
                print(self.cyclePath[i])
            else:
                print(self.cyclePath[i],end="->")
        print()
    

class VisualizeGraph:
    def __init__(self,cyclePath):
        self.cyclePath=cyclePath
        self.graph=nx.DiGraph()
        self.nodes=len(cyclePath)
        self.addEdges()
        self.drawGraph()
    def addEdges(self):
        for i in range(self.nodes-1):
            self.graph.add_edge(self.cyclePath[i],self.cyclePath[(i+1)%self.nodes])
    
    def drawGraph(self):
        pos=nx.spring_layout(self.graph)
        nx.draw(self.graph,pos,with_labels=True,arrows=True)
        plt.show()

if __name__=="__main__":
    cycleDetection=CycleDetectionDFS()
    cycleDetection.runAlgo()
    if cycleDetection.cyclePath:
        visualizeGraph=VisualizeGraph(cycleDetection.cyclePath)
        visualizeGraph.drawGraph()
    else:
        print("No cycle to visualize.")
                
